### ISSUE TITLE
[Proposal] K-Way Merge for Time-Ordered Scans

### ISSUE THREAD

**Issue by justinborromeo** [apache/incubator-druid#7036]
# Motivation

The motivation is to support time-ordering for scan queries that return greater than 100000 rows.  In PR #7024 (which is in review), the ability to time-order small result sets (<100K rows) in memory was added.  It would be nice to have this capability for larger result sets so that SELECT queries can be avoided (time ordering is the only advantage that SELECT has over SCAN).  However, sorting larger datasets in-memory runs the risk of Broker nodes encountering OOMEs.  A different approach needs to be used.  See #6088 for more information.

# Proposed Changes

## Overview

Scan queries return rows in a streaming manner.  Each Historical node executes a query by using a single-threaded sequential scan of target segments.  The query runner streams rows to the client/broker as they're discovered.

## Required changes

To properly implement this feature, a k-way merge would need to occur at both the Historical level (to sort rows from segments) and the Broker level (to sort rows from each Historical and cache) for time-ordered results to be returned to the client.

## Principles

- The change should not affect the performance of non-time-ordered scan queries.
- The performance impact of sorting should be as small as possible.

### Historical-level

Since the existing scan query runner on the Historical is unable to look at all the segments at once, it is impossible to do a k-way merge using the existing query runner.  I propose creating a new query runner factory called something like `OrderedScanQueryRunnerFactory` which would return a single runner when `mergeRunners()` is called.  This runner would open each segment file, perform a k-way merge on the files, and return a `Sequence` of 1-event `ScanResultValues`.  This would involve typecasting the `ScanResultEvent#events` object and splitting the events list into individual events.

Alternatively, to avoid creating a new `QueryRunnerFactory` the existing query runner can check the `ScanQuery` object's limit and ordering, and decide what runner to provide in the `mergeRunners()` function.

As mentioned in #6088, the # of segments queried would have to be capped at some value to prevent the server from running out of memory due to the overhead of opening too many decompression/decoding buffers

### Broker-level

The k-way merge at the broker level should be easier easier to implement in that the existing Sequence.flatMerge() can be used.  The Ordering from the ScanResultValueTimestampComparator introduced in open PR #7024 can be used in the `Sequences#flatMerge()` function.

# Changed Interfaces

- No breaking API changes.
- Returned rows won't be returned with their segmentId (same as in #7024).
- New properties:
    - druid.query.scan.maxSegmentsToTimeOrder
        - Description: The maximum # of segments that a  Historical is allowed to scan in a time-ordered query.  Query fails if number of segments is above this  limit
        - Where: Historical 
        - Default:  TBD
        - Format: int

# Migration / Operational impact

No migration needed.

# Future work

A file-based k-way merge will be needed for large queries that touch more segments than the threshold.

Feel free to let me know if something I wrote is incorrect from a technical point of view.

**Comment by jihoonson** [comment]
Nice proposal! I have some questions/comments.

> This runner would open each segment file, perform a k-way merge on the files, and return a `Sequence` of 1-event `ScanResultValues`.

Does this mean, the merged result in historicals would be a sequence of a single `ScanResultValue` which contains all time-ordered events of input segments? I understand the merge needs to block for sorting, but how does it work with stream merge in the broker? I guess the broker would also wait until it gets all events in `ScanResultValue`?

> Alternatively, to avoid creating a new `QueryRunnerFactory` the existing query runner can check the `ScanQuery` object's limit and ordering, and decide what runner to provide in the `mergeRunners()` function.

I think this is better than adding a new QueryRunnerFactory to minimize the complexity.

**Comment by justinborromeo** [comment]
> Does this mean, the merged result in historicals would be a sequence of a single `ScanResultValue` which contains all time-ordered events of input segments? I understand the merge needs to block for sorting, but how does it work with stream merge in the broker? I guess the broker would also wait until it gets all events in `ScanResultValue`?

Yes.  From `CachingClusteredClient.SpecificQueryRunnable#run()`,

```
return new LazySequence<>(() -> {
        List<Sequence<T>> sequencesByInterval = new ArrayList<>(alreadyCachedResults.size() + segmentsByServer.size());
        addSequencesFromCache(sequencesByInterval, alreadyCachedResults);
        addSequencesFromServer(sequencesByInterval, segmentsByServer);
        return Sequences
            .simple(sequencesByInterval)
            .flatMerge(seq -> seq, query.getResultOrdering());
      });
```

From what I understand, the `Sequences#flatMerge()` function accepts a List of ordered sequences and the ordering for the sequences, then performs a k-way merge.  Since it's assumed that the returned Sequences from the Historicals are already sorted, only the first element of each sequence needs to be materialized to perform the merge.  This should be able to maintain the streaming nature of the Scan query.

**Comment by justinborromeo** [comment]
Adding on to my previous comment, `ScanResultValue`s contain many events at the historical level.  To facilitate streaming and easier sorting using `Sequences#flatMerge()` at the broker level, I plan to "deaggregate" each `ScanResultValue` into several `ScanResultValue`s with one event each.

**Comment by justinborromeo** [comment]
@jihoonson Is it cool to proceed or are there still uncertainties/concerns that still need to be addressed?

**Comment by justinborromeo** [comment]
Closing because it was implemented in #7133
